### DISCUSSION TITLE
Controller defaultValue ignored on <select> tag but work on <input>

### DISCUSSION THREAD

**Discussion by wpsaputra** [react-hook-form/react-hook-form#2815] [Q&A]
I have a problem with following code, 
```
<form>
<Controller
    as={
        <select>
            <option value="" disabled selected>Provinsi...</option>
            {provinsi.map(function (obj) {
                return <option key={obj.kode_prov} value={obj.kode_prov}>{obj.nama_prov}</option>
            })}
        </select>
    }

    control={control}
    name="kode_prov"
    defaultValue={state.data.kode_prov} //---->>>>Not Worked!
    rules={{ required: "Kode provinsi harus terisi" }}
    className={errors.kode_prov ? "error bounce" : ""}

/>
<ErrorMessage errors={errors} name="kode_prov" as="span" />

<input defaultValue={state.data.B1R103} type="text" placeholder="Nomor Kartu Keluarga" name="B1R103" ref={register({ required: "Nomor kartu keluarga harus terisi", 
maxLength: {value : 16, message: "Nomor KK harus 16 digit"}, minLength: {value : 16, message: "Nomor KK harus 16 digit"}, pattern: {value : /^[0-9]*$/, 
message: "Nomor KK hanya boleh mengandung angka"} })} className={errors.B1R103? "error bounce":""} /> //---->>Worked!
<ErrorMessage errors={errors} name="B1R103" as="span" />
</form>
```

I use useEffect hook to fetch api url and set little-state-machine state
`const { action, state } = useStateMachine(updateAction, {shouldReRenderApp: true});`

```
useEffect(() => {
    const abortCtrl = new AbortController();
    const opts = { 
        signal: abortCtrl.signal,
        headers: {
            'Content-Type': 'application/json',
            'X-Authorization': "Bearer "+getAuthData()?.jwt
        } 
    };
	// code to populate select option
	......
	
	//loadBI (get previous selected value by id)
	const loadBI = (opts) => {
        fetch(url_api+'/records/blok_i/'+id, opts)
            .then((response) => response.json())
            .then((data) => {console.log(data); action(data);})
            .catch((error) => console.log(error.message));
    }
    loadBI(opts);
    return () => abortCtrl.abort();
}, []);


```

**Comment by bluebill1049** [comment]
can you reproduce this problem in a simple codesandbox?

**Reply by wpsaputra** [reply]
![desastik](https://user-images.githubusercontent.com/29251295/92195069-b4930980-ee9e-11ea-98ee-b4c12ac13b8b.png)
after execute `action(data)` in useEffect, `<select>` tag defaultValue not changing but `<input>` field is changing, but if we refresh the browser page (f5) all field defaultValue is populated as expected

**Reply by wpsaputra** [reply]
@bluebill1049 Thanks, I will provide the code sandbox asap.

**Reply by bluebill1049** [reply]
is the data comping async? if that's the same, you may need to use `setValue` or `reset` API. `defaultValue` is for the first render.

**Reply by wpsaputra** [reply]
here is the sandbox @bluebill1049 https://codesandbox.io/s/react-hook-form-wizard-form-forked-h8ubo?file=/src/Step1.js when you first open the link, kabupaten field is empty but if you refresh it in codesanbox browser all field appear as intended..

**Reply by wpsaputra** [reply]
To summarize :
kode_kabupaten field : not work at first, only work after refreshing browser
firstName field : work
lastName field : work

My goals is using the same form code for inserting new record and update record. the only difference is in update record I try to set action(data) first by fetch the previous answer from server.

**Reply by sivaputrevu** [reply]
hey @wpsaputra  jus wanna know one thing!
you wanna set a default value for mui select?

**Reply by bluebill1049** [reply]
you shouldn't do this: `const [kabupaten, setKabupaten] = React.useState([]);`

try `reset` / `setValue` API

**Reply by sivaputrevu** [reply]
jus take a look on my example. where I have done default values for single and multiselect.
try render instead of as and rendervalue will get populated with the default values in useForm.
take a look if it can helps.
https://codesandbox.io/s/happy-monad-4lq8i?file=/src/App.tsx

**Reply by bluebill1049** [reply]
Thanks for helping @sivaputrevu ❤️

**Reply by wpsaputra** [reply]
Thanks for the solution @bluebill1049 and @sivaputrevu,  I solved the problem with following code (only for `<Controller>`)   :
`setValue("kode_kab", fetchedPreviousAnswer["kode_kab"], {shouldDirty: true});`

Something that doesn't feel right for me is execute `action(data)` from little-state-machine can update **defaultValue** of `<input>` but not `<Controller>` in react-hook-form regardless of using as or render (in my case` <select>` `<option>` not known beforehand and only known after fetch the API). 

Using `setValue("kode_kab", fetchedPreviousAnswer["kode_kab"], {shouldDirty: true});` is work but feel not clean for me, because some field use defaultValue from little-state-machine directly and the others must use **setValue** manually.

I think the problem could be `action(data)`  from little-state-machine only trigger re-render in `<input>`  but not trigerred in `<Controller>`
